File: go/queue/641_my_circular_deque.go
package queue

// 设计循环双端队列
type MyCircularDeque struct {
	front, rear int
	elements    []int
}

func Constructor3(k int) MyCircularDeque {
	elem := make([]int, k+1)
	return MyCircularDeque{
		elements: elem,
	}
}

// InsertFront 队首插入元素， front = (front -1) % len(elem) 需要考虑front 为负数
func (dq *MyCircularDeque) InsertFront(value int) bool {
	if dq.IsFull() {
		return false
	}
	dq.elements[(dq.front-1+len(dq.elements))%len(dq.elements)] = value
	dq.front = (dq.front - 1 + len(dq.elements)) % len(dq.elements)
	return true
}

func (dq *MyCircularDeque) InsertLast(value int) bool {
	if dq.IsFull() {
		return false
	}

	dq.elements[dq.rear] = value
	dq.rear = (dq.rear + 1) % len(dq.elements)
	return true
}

func (dq *MyCircularDeque) DeleteFront() bool {
	if dq.IsEmpty() {
		return false
	}

	dq.front = (dq.front + 1) % len(dq.elements)
	return true
}

func (dq *MyCircularDeque) DeleteLast() bool {
	if dq.IsEmpty() {
		return false
	}

	dq.rear = (dq.rear - 1 + len(dq.elements)) % len(dq.elements)
	return true
}

func (dq *MyCircularDeque) GetFront() int {
	if dq.IsEmpty() {
		return -1
	}

	return dq.elements[dq.front]
}

func (dq *MyCircularDeque) GetRear() int {
	if dq.IsEmpty() {
		return -1
	}

	return dq.elements[(dq.rear-1+len(dq.elements))%len(dq.elements)]
}

func (dq *MyCircularDeque) IsEmpty() bool {
	return dq.front == dq.rear
}

func (dq *MyCircularDeque) IsFull() bool {
	return (dq.rear+1)%len(dq.elements) == dq.front
}

File: rust/src/queue/641_my_circular_deque.rs

struct MyCircularDeque {
    front: usize,
    rear: usize,
    elements: Vec<i32>,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyCircularDeque {

    fn new(k: i32) -> Self {
        let elements = vec![0; (k+1) as usize];
        MyCircularDeque {front: 0, rear: 0, elements}
    }
    
    fn insert_front(&mut self, value: i32) -> bool {
        if self.is_full() {
            return false
        }

        let n = self.elements.len();
        self.elements[(self.front -1 +n ) % n] = value;

        self.front = (self.front -1 + n) % n;
        true
    }
    
    fn insert_last(&mut self, value: i32) -> bool {
        if self.is_full() {
            return false
        }

        self.elements[self.rear] = value;

        self.rear = (self.rear + 1) % self.elements.len();

        true
    }
    
    fn delete_front(&mut self) -> bool {
        if self.is_empty() {
            return false
        }

        self.front = (self.front + 1) % self.elements.len();

        true
    }
    
    fn delete_last(&mut self) -> bool {
        if self.is_empty() {
            return false
        }

        self.rear = (self.rear -1 + self.elements.len()) % self.elements.len();

        true
    }
    
    fn get_front(&self) -> i32 {
        if self.is_empty() {
            return -1
        }

        self.elements[self.front]
    }
    
    fn get_rear(&self) -> i32 {
        if self.is_empty() {
            return -1
        }
        self.elements[(self.rear-1+self.elements.len()) % self.elements.len()]
    }
    
    fn is_empty(&self) -> bool {
        self.front == self.rear
    }
    
    fn is_full(&self) -> bool {
        self.front == (self.rear + 1) % self.elements.len()
    }
}